pygls/workspace.py

```diff
@@ -80,14 +84,42 @@ def position_from_utf16(lines: List[str], position: Position) -> Position:
     Returns:
         The position with `character` being converted to utf-32 code units.
     """
-    try:
-        return Position(
-            line=position.line,
-            character=position.character
-            - utf16_unit_offset(lines[position.line][:position.character])
+    if len(lines) == 0:
+        return position
+    if position.line <= len(lines) - 1:
+        line = lines[position.line]
+    else:
+        log.warning(
+            f"position_from_utf16() received out of bounds index: lines[{position.line}]"
         )
-    except IndexError:
-        return Position(line=len(lines), character=0)
+        line = lines[-1]
+    if position.character > utf16_num_units(line) - 1:
+        log.warning(
+            f"position_from_utf16() received out of bounds index: line[{position.character}]"
+        )
+        log.debug(f"position_from_utf16() line with {utf16_num_units(line)} UTF16 units: `{line}`")
+        position.character = utf16_num_units(line) - 1
+
```

Comment:
position can be greater than utf16_num_units(line), if for example the cursor is at the end of the line.
I  would just drop this whole block not least because utf16_num_units slows things down.
_apply_incremental_change is designed to work correctly with positions beyond the end of the line.

Reply:
i've kept this just to satisfy this test `assert doc.offset_at_position(Position(line=1, character=5)) == 12`, which seems to make sense to me

Reply:
@pyscripter I need some help to understand why you consider that existing test line to be wrong. Do you disagree with this assertion `assert doc.offset_at_position(Position(line=1, character=5)) == 12`?